KEYCLOAK-16920 Revise OpenShift Guides to meet Red Hat standards (#1266)
keycloak/keycloak-documentation: openshift/topics/advanced_concepts.adoc

> == Performing Advanced Procedures

Reviewer: ```suggestion
== Performing advanced procedures
```

[role="_abstract"]

> This chapter describes advanced procedures, such as seting up keystores and a truststore for the {project_name} server, creating an administrator account, asw well as an overview of available {project_name} client registration methods, and guidance on configuring clustering.

Reviewer: ```suggestion
This chapter describes advanced procedures, such as setting up keystores and a truststore for the {project_name} server, creating an administrator account, as well as an overview of available {project_name} client registration methods, and guidance on configuring clustering.
```

=== Deploying passthrough TLS termination templates

> You can deploy using these templates, which use HTTPS and JGroups keystores and a truststore for the {project_name} server.

Reviewer 2: @andymunro Maybe replace `You can deploy using these templates, which use HTTPS and JGroups keystores and a truststore for the {project_name} server.` with `You can deploy using these templates. They require HTTPS, JGroups keystores and the {project_name} server truststore to already exist, and therefore can be used to instantiate the {project_name} server pod using your custom HTTPS, JGroups keystores and {project_name} server truststore.` ?

The difference compared to your original version is it doesn't say `which use HTTPS, JGroups... etc`. All of the RH-SSO templates, we ship use / require the HTTPS, JGroups keystore and the RH-SSO truststore.

The only difference is:
* For the re-encrypt (x509) ones -- all these keystores & truststore are autogenerated using the OpenShift's x509 service,
* While for the passthrough ones, the user needs to create them manually and provide the location (all the details, like credentials etc) via the environment variables, passed at the moment of template instantiation.

Makes sense?

Author: Okay sounds good.

> You can deploy using these templates, which use HTTPS and JGroups keystores and a truststore for the Red Hat Single Sign-On server.

Reviewer: ```suggestion
You can deploy using these templates, which use HTTPS and JGroups keystores and a truststore for the {project_name} server.
```
Not sure about this one, but maybe it could be replaced?

[[Configuring-Keystores]]
==== Creating HTTPS and JGroups Keystores, and Truststore for the {project_name} Server

> In this procedure, the *_openssl_* toolkit is used to generate a CA certificate to sign the HTTPS keystore, and create a truststore for the {project_name} server. *_keytool_*, a package *included with the Java Development Kit*, is then used to generate self-signed certificates for these keystores.

Reviewer 2: `s/*_keytool_*/The *_keytool_*/` ?

> In this procedure, the *_openssl_* toolkit is used to generate a CA certificate to sign the HTTPS keystore, and create a truststore for the {project_name} server. *_keytool_*, a package *included with the Java Development Kit*, is then used to the generate self-signed certificates for these keystores.

Reviewer: It doesn't look good to me, seems like there some typos, such as "to the generate"?

> The {project_name} application templates using xref:../introduction/introduction.adoc#reencrypt-templates[re-encryption TLS termination] do not *require* or *expect* the HTTPS and JGroups keystores and {project_name} server truststore to be prepared beforehand. The templates use OpenShift's internal link:{ocpdocs_serving_x509_secrets_link}[service serving x509 certificate secrets] to automatically create the HTTPS and JGroups keystores.

Reviewer 2: Maybe here is the right place to add a note / IOW change:
`The {project_name} ... to be prepared beforehand. The templates...` 

to

```
The {project_name} ... to be prepared beforehand. 

Note:
If you want to provision the {project_name} server using existing HTTPS / JGroups keystores, use some of the available passthrough templates instead.

The re-encrypt templates use the OpenShift's internal ....
```

----

[[Configuring-Secrets]]
==== Creating Secrets

.Procedure

You create objects called secrets that OpenShift uses to hold sensitive information, such as passwords or keystores.

. Create the secrets for the HTTPS and JGroups keystores, and {project_name} server truststore, generated in the xref:Configuring-Keystores[previous section].
+
[source,bash,subs="attributes+,macros+"]
----
$ oc create secret generic sso-app-secret --from-file=keystore.jks --from-file=jgroups.jceks --from-file=truststore.jks
----
. Link these secrets to the default service account, which is used to run {project_name} pods.
+
[source,bash,subs="attributes+,macros+"]
----
$ oc secrets link default sso-app-secret
----

[role="_additional-resources"]
.Additional resources
* link:{ocpdocs_secrets_link}[What is a secret?]
* link:{ocpdocs_default_service_accounts_link}[Default project service accounts and roles]

==== Deploying a Passthrough TLS Template using the OpenShift CLI

> After you create xref:Configuring-Keystores[keystores] and xref:Configuring-Secrets[secrets], deploy a passthrough TLS termination template by using the `oc` command.

Reviewer: ```suggestion
.Prerequisites
xref:Configuring-Keystores[Keystores] and xref:Configuring-Secrets[secrets] have been created previosly.

.Procedure
Deploy a passthrough TLS termination template by using the `oc` command.
```
Shouldn't it be something like that?

Author: Yes, that might be good, but I would need a second step to make it a procedure.

Reviewer: Alright, I see, then its fine.

===== `oc` command guidelines
In the following `oc` command, the values of *_SSO_ADMIN_USERNAME_*, *_SSO_ADMIN_PASSWORD_*, *_HTTPS_PASSWORD_*, *_JGROUPS_ENCRYPT_PASSWORD_*, and *_SSO_TRUSTSTORE_PASSWORD_* variables match the default values in the *_{project_templates_version}-https_* {project_name} application template.

For production environments, Red Hat recommends that you consult the on-site policy for your organization for guidance on generating a strong user name and password for the administrator user account of the {project_name} server, and passwords for the HTTPS and JGroups keystores, and the truststore of the {project_name} server.

> Also, when you create the template, make the passwords match the passwords provided when you created the keystores. If you used a different username or password, modify the values of the parameters in your template to match your environment.

Reviewer: As well, this sections seems needs some changes

Reviewer 2: Maybe replace 
`Also, when you create the template, ... to match your environment.` 
with
`Also, if using non-default passwords, make sure the passwords entered as environment variables when instantiating the template match their respective counterparts used to create the HTTPS, JGroups keystores and {project_name_ server truststore.`
?

[NOTE]
====

> You can determine the names associated with the certificate by using the following *_keytool_* commands. *_keytool_* is a package included with the Java Development Kit.

Reviewer 2: Change `names` to `aliases` or `alias names`? (since already changing this, better to be exact)

* *\http://sso-_<project-name>_._<hostname>_/auth/admin*: for the {project_name} web server, and
* *\https://secure-sso-_<project-name>_._<hostname>_/auth/admin*: for the encrypted {project_name} web server.

Use the xref:sso-administrator-setup[administrator user credentials] to log in into the `master` realm's Admin Console.

[[SSO-Clients]]
=== {project_name} clients

Clients are {project_name} entities that request user authentication. A client can be an application requesting {project_name} to provide user authentication, or it can be making requests for access tokens to start services on behalf of an authenticated user. See the link:https://access.redhat.com/documentation/en-us/red_hat_single_sign-on/7.2/html/server_administration_guide/clients[Managing Clients chapter of the {project_name} documentation] for more information.

> {project_name} provides link:{project_doc_base_url}/server_administration_guide/clients#oidc_clients[OpenID-Connect] and link:https://access.redhat.com/documentation/en-us/red_hat_single_sign-on/7.4/html/server_administration_guide/clients#saml_clients[SAML] client protocols.

Reviewer: ```suggestion
{project_name} provides link:{project_doc_base_url}/server_administration_guide/clients#oidc_clients[OpenID-Connect] and link:{project_doc_base_url}/server_administration_guide/clients#saml_clients[SAML] client protocols.
```
Should we update this link?

The following example shows how to set up the files plaintext vault in OpenShift
and set it up to be used for obtaining an SMTP password.

> .Procedure

Reviewer 2: Maybe replace `the files plaintext vault` with `a file-based plaintext vault`?
